File: 백준/Gold/20055. 컨베이어 벨트 위의 로봇/컨베이어 벨트 위의 로봇.java
import java.io.*;
import java.util.*;

public class Main {
	static BufferedReader input = new BufferedReader(new InputStreamReader((System.in)));
	static StringTokenizer tokens;
	static int N, K;
	static int[] belt;
	static boolean[] robot;

	public static void main(String[] args) throws IOException {
		tokens = new StringTokenizer(input.readLine());
		N = Integer.parseInt(tokens.nextToken());
		K = Integer.parseInt(tokens.nextToken());
		belt = new int[2 * N];
		robot = new boolean[2 * N];
		
		tokens = new StringTokenizer(input.readLine());
		for(int i = 0; i < 2 * N; i++) {
			belt[i] = Integer.parseInt(tokens.nextToken());
		}
		
		moveBelt();
	}
	
	private static void moveBelt() {
		int ans = 0;
		
		while(!isEnd()) {
			// 1. 로봇과 벨트 회전
			int tmp = belt[2 * N - 1];
			for(int i = 2 * N -1; i > 0; i--) {
				belt[i] = belt[i-1];
			}
			belt[0] = tmp;

			for(int i = 2 * N - 1; i > 0; i--) {
				robot[i] = robot[i-1];
			}
			robot[0] = false;
			robot[N - 1] = false; // N번 칸 로봇 내리기
			
			// 2. 로봇 이동
			for(int i = 2 * N - 1; i > 0; i--) {
				// 현재 칸에 로봇이 있고, 이동하려는 칸에 로봇이 없고, 칸의 내구도가 1 이상
				if(robot[i-1] && !robot[i] && belt[i] > 0) {
					// 이동하면 칸의 내구도 -1
					belt[i]--;
					robot[i] = true;
					robot[i-1] = false;
				}
			}
			
			robot[N-1] = false; // N번 칸 로봇 내리기
			
			// 3. 로봇 올리기
			if(belt[0] > 0) {
				robot[0] = true;
				belt[0]--;
			}
			
			ans++;
		}
		
		System.out.println(ans);
	}
	
	private static boolean isEnd() {
		int cnt = 0;
		
		for(int i = 0; i < 2 * N; i++) {
			if(belt[i] == 0) {
				cnt++;
			}
		}
		
		if(cnt >= K) {
			return true;
		}
		
		return false;
	}
}

File: 백준/Gold/20055. 컨베이어 벨트 위의 로봇/컨베이어 벨트 위의 로봇.py
import sys
input = sys.stdin.readline

n, k = map(int, input().split())
belt = list(map(int, input().split()))
robot = [0] * n
cnt = 0

while True:    
    # 1
    # 벨트 회전
    belt.insert(0, belt[-1])
    belt = belt[:len(belt)-1]

    # 로봇 회전
    robot.insert(0, robot[-1])
    robot = robot[:len(robot)-1]
    # 내리는 위치에 있는 로봇 없애기
    if robot[-1] == 1:
        robot[-1] = 0

    # 2
    for i in range(len(robot)-1, -1, -1):
        # 로봇 회전
        if robot[i] == 1 and robot[i+1%n] == 0 and belt[i+1%n] > 0:
            robot[i] = 0 # 로봇을 현재 칸에서
            robot[i+1%n] = 1 # 다음 칸으로 이동
            belt[i+1%n] -= 1 # 내구도 감소

    # 내리는 위치면 로봇 없애기
    if robot[-1] == 1:
        robot[-1] = 0

    # 올리는 위치에서 벨트 내구도가 0이 아니면 로봇 올리기
    if belt[0] > 0:
        robot[0] = 1
        belt[0] -= 1

    cnt += 1

    if belt.count(0) >= k:
        break

print(cnt)